### PR TITLE
Improve mAP performance

### LINKED ISSUE #677: Performance issue for own mAP implementation
## 🐛 Bug

The new mAP implementation that replaced the `pycocotools` implementation seems to have a performance issue. In my measurements it is 10-15x slower.

These are are some performance measurements I did on a CPU and single GPU (same machine):

New implementation
```
CPU
Running val metric on 640 samples
Total time: 4.70652174949646
Time per sample 0.007353940233588219

Single GPU
Running val metric on 640 samples
Total time: 33.491135358810425
Time per sample 0.05232989899814129

Running val metric on 64 samples
Total time: 4.064690351486206
Time per sample 0.06351078674197197

Running val metric on 1280 samples
Total time: 61.18194818496704
Time per sample 0.0477983970195055

```

Preview implementation (pycocotools)
```
CPU
Running val metric on 640 samples
Total time: 0.386138916015625
Time per sample 0.000603342056274414

Single GPU
Running val metric on 640 samples
Total time: 2.527284622192383
Time per sample 0.003950447216629982 (bearbeitet) 
```

### To Reproduce

Run mAP implementation as in the example.

#### Code sample

Find the measurement code with BoringModel here:

```python
import time

import torch
import torch.distributed
from pytorch_lightning import LightningModule, Trainer
from torch.utils.data import Dataset, DataLoader

from torchmetrics.detection.map import MAP

BATCH_SIZE = 32

class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


train = RandomDataset(32, 40 * BATCH_SIZE)
train = DataLoader(train, batch_size=BATCH_SIZE)

val = RandomDataset(32, 40 * BATCH_SIZE)
val = DataLoader(val, batch_size=BATCH_SIZE)

# mockups for MAP compatible data
mock_preds = [
    dict(
        boxes=torch.Tensor([[258.0, 41.0, 606.0, 285.0]]),
        scores=torch.Tensor([0.536]),
        labels=torch.IntTensor([0]),
    )
]
mock_target = [
    dict(
        boxes=torch.Tensor([[214.0, 41.0, 562.0, 285.0]]),
        labels=torch.IntTensor([0]),
    )
]


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.val_map = MAP(class_metrics=True, dist_sync_on_step=True)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)

        # ignore real outputs and add mockup preds to metric
        preds = []
        target = []

        for n in range(batch.size(0)):
            x = mock_preds[0]
            preds.append({
                'boxes': x['boxes'].to(self.device),
                'labels': x['labels'].to(self.device),
                'scores': x['scores'].to(self.device)
            })

            x = mock_target[0]
            target.append({
                'boxes': x['boxes'].to(self.device),
                'labels': x['labels'].to(self.device)
            })

        self.val_map.update(preds=preds, target=target)
        return {"x": loss}

    def on_validation_epoch_start(self) -> None:
        self.val_map.reset()

    def on_validation_epoch_end(self) -> None:
        if self.trainer.global_step != 0:
            print(
                f"\nRunning val metric on {len(self.val_map.groundtruth_boxes)} samples"
            )
            start = time.time()
            result = self.val_map.compute()  # GPUs get stuck here
            end = time.time()
            print(f"Total time: {end-start}")
            print(f"Time per sample {(time.time() - start) / len(self.val_map.groundtruth_boxes)}")

    def configure_optimizers(self):
        return [torch.optim.SGD(self.layer.parameters(), lr=0.1)]


model = BoringModel()
trainer = Trainer(
    max_epochs=10,
    strategy='ddp',
    gpus=1
)

trainer.fit(model, train, val)
```

### Expected behavior

Performance should at least be on par with the pycocotools implementation.

Best case, it should be faster, especially on GPU.

### Environment

- PyTorch Version (e.g., 1.0): 1.0
- OS (e.g., Linux): Windows 10
- How you installed PyTorch (`conda`, `pip`, source): pip
- Build command you used (if compiling from source):
- Python version: 3.9
- CUDA/cuDNN version: 11.3
- GPU models and configuration: GTX 1060
- Any other relevant information:

### PR DESCRIPTION
**Pull request by twsl**
## What does this PR do?

Fixes #677

Work of @tkupek @OlofHarrysson  @twsl 
First steps to get performance on par with pycocotools/numpy

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tkupek** [comment]
The performance improvements are great and really necessary.

However, CUDA calculations are still by a factor 9-10 slower than CPU. Computations are not CUDA optimized and I don't know if it's possible (or how much effort it is).

```
Running metric on 10 samples on cuda:0
Total time: 8.060346603393555
Time per sample 0.8060346603393554

Running metric on 10 samples on cpu
Total time: 0.98065185546875
Time per sample 0.098065185546875
```

Until then, I suggest to **make the mAP metric CPU only**, to avoid that somebody uses the slow CUDA version.
I would move everything to CPU, as we did it with the `pycocotools` implementation.

@twsl @Borda @OlofHarrysson @SkafteNicki

**Comment by Borda** [comment]
> Until then, I suggest to **make the mAP metric CPU only**, to avoid that somebody uses the slow CUDA version.
> I would move everything to CPU, as we did it with the `pycocotools` implementation.

sounds reasonable to me :rabbit:

File: torchmetrics/detection/map.py [hunk 1]
```diff
@@ -456,43 +452,16 @@ def _evaluate_image(
...
-            if match_id > -1 and not gt_ignore[match_id] and gt_ignore[idx_gt]:
-                break
-            # continue to next gt unless better match made
-            if ious[idx_det, idx_gt] < iou:
-                continue
-            # if match successful and best so far, store appropriately
-            iou = ious[idx_det, idx_gt]
-            match_id = idx_gt
-        return match_id
+        previously_matched = gt_matches[idx_iou]
```
Comment (SkafteNicki) [inline torchmetrics/detection/map.py:457]: could we reinsert the docstring here?

File: torchmetrics/detection/map.py [hunk 2]
```diff
@@ -570,12 +541,12 @@ def _calculate(self, class_ids: List) -> Tuple[Dict, MAPMetricResults, MARMetric
         scores = -torch.ones((nb_iou_thrs, nb_rec_thrs, nb_classes, nb_bbox_areas, nb_max_det_thrs))
 
         # move tensors if necessary
-        self.max_detection_thresholds = self.max_detection_thresholds.to(self.device)
-        self.rec_thresholds = self.rec_thresholds.to(self.device)
+        # self.max_detection_thresholds = self.max_detection_thresholds.to(self.device)
```
Comment (SkafteNicki) [inline torchmetrics/detection/map.py:544]:
```suggestion
```

File: torchmetrics/detection/map.py [hunk 3]
```diff
@@ -425,22 +421,22 @@ def _evaluate_image(
         nb_iou_thrs = len(self.iou_thresholds)
         nb_gt = len(gt)
         nb_det = len(det)
-        gt_matches = torch.zeros((nb_iou_thrs, nb_gt), dtype=torch.bool, device=self.device)
-        det_matches = torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool, device=self.device)
+        gt_matches = torch.zeros((nb_iou_thrs, nb_gt), dtype=torch.bool)
+        det_matches = torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool)
         gt_ignore = ignore_area_sorted
-        det_ignore = torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool, device=self.device)
+        det_ignore = torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool)
```
Comment (SkafteNicki) [inline torchmetrics/detection/map.py:427]: is it really safe to remove all these `device=...`?

File: torchmetrics/detection/map.py [hunk 4]
```diff
@@ -666,19 +642,17 @@ def __calculate_recall_precision_scores(
             recall[idx, idx_cls, idx_bbox_area, idx_max_det_thrs] = rc[-1] if nd else 0
 
             # Remove zigzags for AUC
-            for i in range(nd - 1, 0, -1):
-                if pr[i] > pr[i - 1]:
-                    pr[i - 1] = pr[i]
+            diff_zero = torch.zeros((1,))
+            diff = torch.ones((1,))
```
Comment (SkafteNicki) [inline torchmetrics/detection/map.py:646]:
```suggestion
            diff_zero = torch.zeros(1,)
            diff = torch.ones(1,)
```

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
@tkupek @OlofHarrysson @twsl thanks for really trying to improve performance for this metric.
I agree that we should merge this and do a small release.
However, out hole GPU testing pipeline is down at the moment so no PRs are currently getting merged :(

File: torchmetrics/detection/map.py (hunk 3 above)
Comment (tkupek) [reply torchmetrics/detection/map.py:427]: Will be instantiated on the CPU which is exactly what we need.

Comment (twsl) [reply torchmetrics/detection/map.py:427]: I had the same thoughts. Adding these improve performance significantly, if run on gpu. but keeping all on the cpu is still a lot faster.

File: torchmetrics/detection/map.py [hunk 5]
```diff
@@ -664,19 +663,17 @@ def __calculate_recall_precision_scores(
...
+            while not torch.all(diff == 0):
+                diff = torch.clamp(torch.cat((torch.diff(pr), diff_zero), 0), min=0)
```
Comment (SkafteNicki) [inline torchmetrics/detection/map.py:669]: `torch.diff` was first introduced in v1.8. This solution is better for backwards compatibility.
```suggestion
                diff = torch.clamp(torch.cat((pr[1:] - pr[:-1], diff_zero), 0), min=0)
```

**Comment by SkafteNicki** [comment]
@Borda please approve and merge.
Passes locally for multi-gpu support:
image

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
> Pycocotools (previous implementation): 74.91s 
> Current implementation: 1742.07s
> This branch: 140.79s
>
> Further improvements should be done to get closer to the pycocotools benchmark.

@twsl @tkupek do you think that you can make further improvements to get speed parity with cocotooks?

**Comment by tkupek** [comment]
Unfortunately, I don't have any concrete ideas in mind, nor the time to look at it right now.

**Comment by twsl** [comment]
I might give it another shot after i handed in my thesis next week. But cant promise any results or timeline
